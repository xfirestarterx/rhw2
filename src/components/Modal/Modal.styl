@import '../../variables.styl'

.ModalWrap
  position fixed
  top 0
  right 0
  bottom 0
  left 0
  z-index 1
  display flex
  justify-content center
  align-items center

.Overlay
  position fixed
  top 0
  right 0
  bottom 0
  left 0
  background rgba(0,0,0,.5)
  z-index 2

.ModalContent
  position relative
  z-index 3
  padding 40px
  background $dark300
  width 50vw
  max-height 90vh
  overflow-y auto

.ModalTitle
  text-transform uppercase
  font-size 21px
  margin-top 0

.ModalClose
  width 28px
  height 28px
  position absolute
  top 20px
  right 20px
  background url('../../img/modal-close.png') no-repeat 0 0
  cursor pointer
