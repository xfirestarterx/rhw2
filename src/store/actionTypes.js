export const OPEN_MODAL = 'OPEN_MODAL';
export const CLOSE_MODAL = 'CLOSE_MODAL';
export const SET_MOVIES = 'SET_MOVIES';
export const DELETE_MOVIE = 'DELETE_MOVIE';
export const SET_IS_LOADING = 'SET_IS_LOADING';
