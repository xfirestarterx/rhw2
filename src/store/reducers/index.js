import { combineReducers } from 'redux';
import modal from './modal';
import movies from './movies';

export default combineReducers({ modal, movies });
