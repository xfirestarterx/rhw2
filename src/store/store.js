import { createStore } from 'redux';
import rootReducer from './reducers/index';
import { composeWithDevTools } from 'redux-devtools-extension';

export default createStore(rootReducer, composeWithDevTools());
