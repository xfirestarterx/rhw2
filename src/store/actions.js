import * as actionTypes from './actionTypes';
import * as moviesSvc from '../utils/moviesSvc';
import normalizeMoviesData from '../utils/normalizeMoviesData';
import setQueryParams from '../utils/setQueryParams';
import store from './store';

export const openModal = (modalType, movieId) => ({
  type: actionTypes.OPEN_MODAL,
  payload: {
    currentModal: modalType,
    movieId
  }
});

export const closeModal = () => ({
  type: actionTypes.CLOSE_MODAL
});

const setMovies = (movies) => ({
  type: actionTypes.SET_MOVIES,
  payload: {
    movies
  }
});

const updateRequestState = (terms) => ({
  type: actionTypes.UPDATE_REQUEST_PARAMS,
  payload: terms
});

export const setMoviesByTerms = (terms) => {
  const requestParamsState = store.getState().movies.params;
  const joinedTerms = {
    ...requestParamsState,
    ...terms
  }
  
  return async (dispatch) => {
    dispatch(updateRequestState(terms));
    if (!joinedTerms.search) return;

    setQueryParams(joinedTerms);

    dispatch(setIsLoading(true));

    const movies = await moviesSvc.getByTerms(joinedTerms);
    dispatch(setMovies(normalizeMoviesData(movies)));

    dispatch(setIsLoading(false));
  }
}

const setMovieDetails = (movie) => ({
  type: actionTypes.GET_MOVIE_DETAILS,
  payload: movie
});

export const setMovieDetailsThunk = (id) => {
  return async (dispatch) => {
    const movie = await moviesSvc.get(id);
    dispatch(setMovieDetails(normalizeMoviesData([movie])));
  }
}

const addMovie = movie => ({
  type: actionTypes.ADD_MOVIE,
  payload: {
    movie
  }
});

export const addMovieThunk = movie => {
  return async (dispatch) => {
    await moviesSvc.add(movie);
    dispatch(addMovie(normalizeMoviesData([movie])));
  }
}

const editMovie = movie => ({
  type: actionTypes.EDIT_MOVIE,
  payload: {
    movie
  }
});

export const editMovieThunk = movie => {
  return async (dispatch) => {
    await moviesSvc.update(movie);
    dispatch(editMovie(normalizeMoviesData([movie])));
  }
}

const deleteMovie = id => ({
  type: actionTypes.DELETE_MOVIE,
  payload: {
    id
  }
});

export const deleteMovieThunk = id => {
  return async (dispatch) => {
    await moviesSvc.remove(id);
    dispatch(deleteMovie(id));
  }
}

export const setIsLoading = (isLoading) => ({
  type: actionTypes.SET_IS_LOADING,
  payload: {
    isLoading
  }
});
