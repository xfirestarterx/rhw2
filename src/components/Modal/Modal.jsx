import React from 'react';
import styles from './Modal.styl';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { closeModal } from 'store/actions';

export const modalType = {
  add: 'add',
  delete: 'delete',
  edit: 'edit',
  none: 'none'
}

const Modal = ({ title, children, closeModal }) => {
  return (
    <div className={styles.ModalWrap}>
      <div onClick={closeModal} className={styles.Overlay} />
      <div className={styles.ModalContent}>
        <div onClick={closeModal} className={styles.ModalClose} />
        { title && <p className={styles.ModalTitle}>{title}</p> }
        { children }
      </div>
    </div>
  );
};

Modal.propTypes = {
  title: PropTypes.string,
  children: PropTypes.any
}

export default connect(null, { closeModal })(Modal);

