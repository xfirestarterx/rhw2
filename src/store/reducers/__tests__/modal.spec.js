import store from 'store/store';
import { openModal, closeModal } from 'store/actions';
import { modalType } from 'components/Modal/Modal';
import reducer from '../modal';

describe('Modal reducer', () => {
  const movieId = '1';

  it('returns the initial state', () => {
    expect(reducer(undefined, {})).toMatchSnapshot();
  });

  it('returns correct state when opening modal', () => {
    const expectedState = {
      currentModal: modalType.edit,
      isModalShown: true,
      movieId,
    };

    expect(
      reducer(undefined, openModal(modalType.edit, movieId)),
    ).toEqual(expectedState);
  });

  it('returns correct state when closing modal', () => {
    const expectedState = {
      isModalShown: false,
      currentModal: modalType.none,
      movieId: undefined,
    };

    expect(
      reducer(undefined, closeModal()),
    ).toEqual(expectedState);
  });
});
