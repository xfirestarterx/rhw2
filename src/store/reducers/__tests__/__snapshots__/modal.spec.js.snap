// Jest Snapshot v1, https://goo.gl/fbAQLP

exports['Modal reducer returns the initial state 1'] = `
Object {
  "currentModal": "none",
  "isModalShown": false,
  "movieId": undefined,
}
`;
